src/agent/src/mount.rs

```diff
@@ -51,6 +52,75 @@ const RO_MASK: u32 = 0o440;
 const EXEC_MASK: u32 = 0o110;
 const MODE_SETGID: u32 = 0o2000;
 
+pub trait RefCounted {
+    // Increment the reference count by one
+    fn inc_ref(&self);
+
+    // Decrement the reference count by one
+    fn dec_ref(&self) -> Result<()>;
+
+    // Return true if the reference count is zero
+    fn is_zero(&self) -> bool;
+
+    //Return the count number
+    fn get_count(&self) -> u32;
+}
+#[derive(Debug)]
+pub struct RefCountedObject<T> {
+    pub name: String,
+    pub ref_count: AtomicU32,
+    pub ref_type: T,
+}
+
+// Implement the RefCounted trait for the generic struct
+impl<T> RefCounted for RefCountedObject<T> {
+    fn inc_ref(&self) {
+        self.ref_count.fetch_add(1, Ordering::Relaxed);
+    }
+
+    fn dec_ref(&self) -> Result<()> {
+        self.ref_count.fetch_sub(1, Ordering::Relaxed);
+        Ok(())
+    }
+
+    fn is_zero(&self) -> bool {
+        self.ref_count.load(Ordering::Relaxed) == 0
+    }
+
+    fn get_count(&self) -> u32 {
+        self.ref_count.load(Ordering::Relaxed)
+    }
+}
+#[derive(Debug)]
+// Define an enum to represent the MountPoint and VerityDevice variants
+pub enum ObjectKind {
+    MountPoint,
+    VerityDevice,
+}
+
+// Implement a custom dec_ref method for each variant
+impl RefCountedObject<ObjectKind> {
+    fn dec_ref(&self) -> Result<()> {
+        self.ref_count.fetch_sub(1, Ordering::Relaxed);
+        match self.ref_type {
+            ObjectKind::MountPoint => {
+                let mounts = vec![self.name.clone()];
+                remove_mounts(&mounts)?;
```

Comment:
Do you really want to remove mounts every time the refcount is decremented? Shouldn't you check that it's zero before removing?

Reply:
You're right. I will change the codes